upgrade/v102to110/upgrade.go

```diff
@@ -0,0 +1,129 @@
+// should be updated, too.
+// Link to the original issue: https://github.com/longhorn/longhorn/issues/1269
+
+const (
+	upgradeLogPrefix = "upgrade from v1.0.2 to v1.1.0: "
+)
+
+func UpgradeCRDs(namespace string, lhClient *lhclientset.Clientset) error {
+	if err := createDisksAndUpdateNodes(namespace, lhClient); err != nil {
+		return err
+	}
+	if err := doReplicaUpgrade(namespace, lhClient); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createDisksAndUpdateNodes(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade instance manager failed")
+	}()
+
+	nodeList, err := lhClient.LonghornV1beta1().Nodes(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, upgradeLogPrefix+"failed to list all existing nodes during the upgrade")
+	}
+
+	for _, node := range nodeList.Items {
+		if node.Spec.Disks == nil {
+			continue
+		}
+		for diskID, diskSpec := range node.Spec.Disks {
+			// Disks on different nodes may be using the same disk ID.
+			// And the generated name for each node disk during the upgrade is fixed.
+			diskName := generateFixedDiskNameForExistingDisks(node.Name, diskID)
+			if _, err := lhClient.LonghornV1beta1().Disks(namespace).Get(diskName, metav1.GetOptions{}); err == nil {
+				continue
+			} else {
+				if !apierrors.IsNotFound(err) {
+					return errors.Wrapf(err, upgradeLogPrefix+"failed to get disk %v for node %v during the upgrade", diskName, node.Name)
+				}
+			}
+			disk := &longhorn.Disk{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      diskName,
+					Namespace: node.Namespace,
+					Labels: map[string]string{
+						types.LonghornNodeKey: node.Name,
+					},
+				},
+				Spec: types.DiskSpec{
+					NodeID:            node.Name,
+					Path:              diskSpec.Path,
+					AllowScheduling:   diskSpec.AllowScheduling,
+					EvictionRequested: diskSpec.EvictionRequested,
+					StorageReserved:   diskSpec.StorageReserved,
+					Tags:              diskSpec.Tags,
+				},
+			}
+			if _, err := lhClient.LonghornV1beta1().Disks(namespace).Create(disk); err != nil {
+				return errors.Wrapf(err, upgradeLogPrefix+"failed to create disk %v during the upgrade", diskName)
+			}
+		}
+
+		node.Spec.Disks = nil
+		node.Status.DiskStatus = nil
+		if _, err := lhClient.LonghornV1beta1().Nodes(namespace).Update(&node); err != nil {
+			return errors.Wrapf(err, upgradeLogPrefix+"failed to clean up the deprecated field node.Spec.Disks for node %v during the upgrade", node.Name)
+		}
+	}
+	return nil
+}
+
+func getNodeLabelSelector(nodeName string) string {
+	return fmt.Sprintf("%s=%s", types.LonghornNodeKey, nodeName)
+}
+
+func generateFixedDiskNameForExistingDisks(nodeName, diskID string) string {
+	return fmt.Sprintf("%s-%s", diskID, util.GetStringChecksum(strings.TrimSpace(nodeName))[:8])
+}
```

Comment:
@shuo-wu This function is problematic when there is a fail-to-schedule replica in the cluster.  In this case, the diskID  of the  fail-to-schedule replica is empty, so the return value will have format `-xxxxxxxx`. That is invalid for Kubernetes label value.  Please see this error for more details:
```
time="2020-09-23T07:39:23Z" level=error msg="Upgrade failed: upgrade CRD failed: upgrade from v1.0.2 to v1.1.0: upgrade instance manager failed: Replica.longhorn.io \"testvol3-r-12d412a9\" is invalid: metadata.labels: Invalid value: \"-cf83e135\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')"
```

Reply:
This implementation has been deprecated. And thank you for pointing out the error!